File: Source/ViewControllerPresentationSpy/UIAlertAction+QCOMock.m
//  ViewControllerPresentationSpy by Jon Reid, https://qualitycoding.org
//  Copyright 2022 Jonathan M. Reid. See LICENSE.txt

#import "UIAlertAction+QCOMock.h"

#import "NSObject+QCOMockAlerts.h"
#import "ViewControllerPresentationSpy/ViewControllerPresentationSpy-Swift.h"
#import <objc/runtime.h>

void *const foo = @"foo";

@implementation UIAlertAction (QCOMock)

+ (instancetype)qcoMock_actionWithTitle:(NSString *)title
                                  style:(UIAlertActionStyle)style
                                handler:(void (^ __nullable)(UIAlertAction *action))handler
{
    UIAlertAction *action = [self qcoMock_actionWithTitle:title style:style handler:handler];
    UIAlertActionExtraProperties *extraProperties = [[UIAlertActionExtraProperties alloc] initWithHandler:handler];
    objc_setAssociatedObject(action, foo, extraProperties, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
    return action;
}

- (void (^ __nullable)(UIAlertAction *action))qcoMock_handler
{
    UIAlertActionExtraProperties *extraProperties = objc_getAssociatedObject(self, foo);
    return extraProperties.handler;
}

@end

File: Source/ViewControllerPresentationSpy/UIAlertAction+QCOMock.swift
//  ViewControllerPresentationSpy by Jon Reid, https://qualitycoding.org
//  Copyright 2022 Jonathan M. Reid. See LICENSE.txt

import UIKit

extension UIAlertAction {
    @objc class func qcoMock_swizzle() {
        UIAlertAction.qcoMockAlerts_replaceClassMethod(
            #selector(UIAlertAction.init(title:style:handler:)),
            withMethod: #selector(UIAlertAction.qcoMock_action(withTitle:style:handler:))
        )
    }

    @objc class func qcoMock_action(withTitle title: String, style: UIAlertAction.Style, handler: ((UIAlertAction) -> Void)?) -> UIAlertAction {
        let action = UIAlertAction.qcoMock_action(withTitle: title, style: style, handler: handler)
        let extraProperties = UIAlertActionExtraProperties(handler: handler)
        objc_setAssociatedObject(action, foo, extraProperties, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        return action
    }
}

public final class UIAlertActionExtraProperties: NSObject {
    @objc public static let associatedObjectKey = "extraProperties"

    @objc public let handler: ((UIAlertAction) -> Void)?

    init(handler: ((UIAlertAction) -> Void)?) {
        self.handler = handler
        super.init()
    }
}
